esphome/components/mcp9808/mcp9808.cpp

```diff
@@ -0,0 +1,76 @@
+#include "mcp9808.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace mcp9808 {
+
+static const uint8_t MCP9808_REG_AMBIENT_TEMP = 0x05;
+static const uint8_t MCP9808_REG_MANUF_ID = 0x06;
+static const uint8_t MCP9808_REG_DEVICE_ID = 0x07;
+
+static const char *TAG = "mcp9808";
+
+void MCP9808Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up MCP9808...");
+
+  uint16_t manu;
+  if (!this->read_byte_16(MCP9808_REG_MANUF_ID, &manu, 0) || manu != 0x0054) {
+    this->mark_failed();
+    ESP_LOGCONFIG(TAG, "MCP9808 manufacuturer id failed");
+    return;
+  }
+  uint16_t dev_id;
+  if (!this->read_byte_16(MCP9808_REG_DEVICE_ID, &dev_id, 0) || dev_id != 0x0400) {
+    this->mark_failed();
+    ESP_LOGCONFIG(TAG, "MCP9808 device id failed");
+    return;
+  }
+}
+void MCP9808Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "MCP9808:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with MCP9808 failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Temperature", this->temperature_);
+}
+void MCP9808Component::update() {
+  uint16_t raw_temp;
+  if (!this->read_byte_16(MCP9808_REG_AMBIENT_TEMP, &raw_temp)) {
+    this->status_set_warning();
+    return;
+  }
+  if (raw_temp == 0xFFFF) {
+    this->status_set_warning();
+    return;
+  }
+
+  // http://ww1.microchip.com/downloads/en/DeviceDoc/25095A.pdf
+
+  float temp = NAN;
+  uint8_t msb = (uint8_t)((raw_temp & 0xff00) >> 8);
```

Comment:
In C++ code, I would use C++ style casts
```suggestion
  uint8_t msb = static_cast<uint8_t>((raw_temp & 0xff00) >> 8);
```

Reply:
@pkuehne In pure C++ project's I'd agree, but for ESPHome it's fine. We can't expect all contributors to know C++ perfectly, especially when most do not come from C++ backgrounds.